third_party/USB-UIRT/Transmit/uu_irsage.cpp

```diff
@@ -0,0 +1,1008 @@
+{
+  char *errStr = NULL;
+
+  char devName[256];
+
+  DbgLog("OpenDevice: dev #%d\n",deviceNum);
+
+  if ((deviceNum < 0) || (deviceNum > 7))
+    return false;
+
+  if ((hDrvHandle[deviceNum]) && (hDrvHandle[deviceNum] != INVALID_HANDLE_VALUE))
+    return true;
+
+  while (1)
+  {
+    if (!loadDLL())
+    {
+      errStr = "Error: Unable to find USB-UIRT Driver. Please make sure driver is Installed!";
+      break;
+    }
+
+    if (!fn_UUIRTGetDrvInfo(&drvVersion))
+    {
+      errStr = "Error: Unable to retrieve USB-UIRT Driver Version!";
+      unLoadDLL();
+      break;
+    }
+
+    if ((!fn_UUIRTGetDrvInfo(&drvVersion)) || (drvVersion != 0x0100))
+    {
+      errStr = "Error: USB-UIRT Driver Version (uuirtdrv.dll) is incompatible with this plugin. Please make sure you are using the latest version of plugin and driver!";
+      unLoadDLL();
+      break;
+    }
+
+    dllVersion = 0;
+    if (fnUUIRTOpenEx)
+      fn_UUIRTGetDrvVersion(&dllVersion);
+    if ((!fnUUIRTOpenEx) || (dllVersion < ((deviceNum > 0) ? (unsigned) 2670 : (unsigned) 2640)))
+    {
+      errStr = "Error: USB-UIRT Driver Version (uuirtdrv.dll) is incompatible with this plugin. Please make sure you are using version 2.6.4 or later of the USB-UIRT API driver!";
+      MessageBox(0, errStr, "Error", MB_OK);
+      unLoadDLL();
+      break;
+    }
+
+
+    if (deviceNum)
+      sprintf_s(devName, 256, "USB-UIRT-%d", deviceNum+1);
+    else
+      strcpy_s(devName, 256, "USB-UIRT");
+
+    hDrvHandle[deviceNum] = fnUUIRTOpenEx(devName, 0, NULL, NULL);
+    if (hDrvHandle[deviceNum] == INVALID_HANDLE_VALUE)
+    {
+      DWORD err;
+
+      err = GetLastError();
+
+      if (err == UUIRTDRV_ERR_NO_DLL)
+      {
+        errStr = "Error: Unable to find USB-UIRT Driver. Please make sure driver is Installed!";
+      }
+      else if (err == UUIRTDRV_ERR_NO_DEVICE)
+      {
+        errStr = "Error: Unable to connect to USB-UIRT device!  Please ensure device is connected to the computer!\n";
+      }
+      else if (err == UUIRTDRV_ERR_NO_RESP)
+      {
+        errStr = "Error: Unable to communicate with USB-UIRT device!  Please check connections and try again.  If you still have problems, try unplugging and reconnecting your USB-UIRT.  If problem persists, contact Technical Support!\n";
+      }
+      else if (err == UUIRTDRV_ERR_VERSION)
+      {
+        errStr = "Error: Your USB-UIRT's firmware is not compatible with the USB-UIRT driver installed on this machine. Please verify you are running the latest API DLL and that you're using the latest version of USB-UIRT firmware!  If problem persists, contact Technical Support!\n";
+      }
+      else
+      {
+        static char myStr[256];
+
+        sprintf_s(myStr,256, "Error: Unable to initialize USB-UIRT (unknown error) [%08lx]!\n",err);
+        DbgLog(myStr);
+        errStr = myStr;
+      }
+
+      unLoadDLL();
+
+      break;
+    }
+
+    break;
+  }
+
+  if (errStr)
+  {
+    return false;
```

Comment:
Maybe log the errStr? It just seems weird there's so much effort to craft different ones above that are never utilized.

Reply:
This file came from Jon @ USB-UIRT and I actually added the log file logging.  Previously the string went to a commented out MessageBox() function which obviously only works in a Windows app.  I actually added the file logging option when debugging.  I'll include the errStr since it makes sense to do so.